metrics/latency.go

```diff
@@ -82,3 +82,120 @@ func ProbeLatency(ch channel2.Channel, histogram Histogram, interval time.Durati
 		}
 	}
 }
+
+type LatencyHandler interface {
+	LatencyReported(latency time.Duration)
+	ChannelClosed()
+}
+
+func AddLatencyProbe(ch channel2.Channel, interval time.Duration, handler LatencyHandler) {
+	probe := &latencyProbe{
+		handler:  handler,
+		ch:       ch,
+		interval: interval,
+	}
+	ch.AddReceiveHandler(probe)
+	go probe.run()
+}
+
+type latencyProbe struct {
+	handler  LatencyHandler
+	ch       channel2.Channel
+	interval time.Duration
+}
+
+func (self *latencyProbe) ContentType() int32 {
+	return channel2.ContentTypeLatencyResponseType
+}
+
+func (self *latencyProbe) HandleReceive(m *channel2.Message, _ channel2.Channel) {
+	if sentTime, ok := m.GetUint64Header(latencyProbeTime); ok {
+		latency := time.Duration(time.Now().UnixNano() - int64(sentTime))
+		self.handler.LatencyReported(latency)
+	} else {
+		pfxlog.Logger().Error("no send time on latency response")
+	}
+}
+
+func (self *latencyProbe) run() {
+	log := pfxlog.ContextLogger(self.ch.Label())
+	log.Debug("started")
+	defer log.Debug("exited")
+	defer self.handler.ChannelClosed()
+
+	lastSend := info.NowInMilliseconds()
+	for {
+		time.Sleep(self.interval)
+		if self.ch.IsClosed() {
+			return
+		}
+
+		now := info.NowInMilliseconds()
+		if now-lastSend > 10000 {
```

Comment:
Can we make this configurable? Probably should fix up any hard-coded timeouts anywhere we run across them (especially in new code).

Reply:
I just removed this check. I had copied this from the original latency code, but really, there's no reason to double check the latency interval that's getting passed in. The value that's passed in is settable via config.